[Optimize] Move dead partial apply elimination to mandatory combine

zoecarver opened the pull request:
This patch moves the removal of dead partial applies to mandatory combine (from mandatory inlining). SourceKit doesn't run mandatory combine but, it does run mandatory inlining, so the more we can move out of mandatory inlining, the better. 

refs #28484 #28394 #28396
cc @gottesmm

zoecarver committed:
Move partial_apply closure cleanup into mandatory combine

zoecarver committed:
Add comments

zoecarver committed:
Format

zoecarver commented:
@gottesmm friendly ping. Does the general idea of this patch look good to you?

slavapestov commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -270,6 +271,23 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
+  // Remove the stack dealloc, then the partial apply, then the function ref.
+  instModCallbacks.deleteInst(user->getUser());
+  instModCallbacks.deleteInst(i);
+  instModCallbacks.deleteInst(cast<FunctionRefInst>(i->getCallee()));
```
It's unlikely but possible that the function_ref has other uses, so I'm not sure it's quite safe to remove it.

zoecarver replied:
Good point. I'll fix this.

slavapestov reviewed (commented):
Are there any tests that exercise the old code? If not, do you mind adding one?

slavapestov commented:
@swift-ci Please smoke test

zoecarver commented:
@slavapestov this should mostly be a NFC. So, no tests should need to change. Do you mean I should add a test that only runs mandatory inlining (to make sure the closure isn't optimized)?

zoecarver committed:
Merge branch 'master' into optimize/dead-closure-in-mand-combine

zoecarver committed:
Fix failing tests

slavapestov commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -270,6 +271,25 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
+  instModCallbacks.deleteInst(i);
+  // If the only use of the function_ref is us, then remove it.
+  if (i->getCallee()->getSingleUse()->getUser() == i)
+    instModCallbacks.deleteInst(cast<FunctionRefInst>(i->getCallee()));
```
I don't think you can guarantee that the callee is always going to be a function_ref. Can you add an isa check to make sure?

slavapestov commented:
It's up to you if you want to add a SIL test that only runs mandatory combining or not.

slavapestov commented:
@swift-ci Please smoke test

zoecarver committed:
Merge branch 'master' into optimize/dead-closure-in-mand-combine

zoecarver committed:
Fix mandatory_inlining.sil

zoecarver committed:
Rename tests

zoecarver commented on test/SILOptimizer/mandatory_inlining_and_combine.sil:
```diff
@@ -0,0 +1,317 @@
+  return %5 : $Builtin.Int8
+}
+
+sil shared [transparent] @identity_closure : $@convention(thin) (Builtin.Int8) -> Builtin.Int8 {
```
I'm ignoring the fact that this function no longer gets removed because it seems inline with the update in #28484

zoecarver commented on test/SILOptimizer/mandatory_inlining.sil:
```diff
@@ -1088,7 +1012,7 @@ bb0(%0 : $C):
 // CHECK:   apply [[F]]([[ARG]])
 // CHECK:   strong_release [[ARG]]
-// CHECK:   strong_release [[ARG]]
+// TODO: why is this broken?
```
I need to figure out why there are too few `strong_release`s here.

zoecarver committed:
stash: add load visitor and refactor

zoecarver committed:
Use lambda to remove instructions

zoecarver committed:
Move stripCopiesAndBorrows and cleanupLoadedCalleeValue into InstOptUtils

zoecarver committed:
Fix crash and remove todos

zoecarver commented on test/SILOptimizer/mandatory_inlining_and_combine.sil:
```diff
@@ -0,0 +1,317 @@
+// CHECK-NEXT:  br [[MERGE_BB]]
+//
+// CHECK: [[MERGE_BB]]:
+// TODO: strong_release [[ARG]]
```
I was able to fix the other `strong_release` todos but I'm still looking into the cause of this one.

zoecarver commented:
@slavapestov I've added `mandatory_inlining_and_combine.sil` to test that mandatory inlining and mandatory combine each do the "right" thing. 

I had to update this patch quite a bit to cover all the closure cleanup done in mandatory inlining, now it should cover all the cases that it used to. 

`mandatory_inlining_ownership2.sil` had a lot of tests that now required both a mandatory inlining and mandatory combine pass. Instead of moving them to `mandatory_inlining_and_combine.sil` I added a `RUN` comment that runs both passes. If you'd prefer, I'm happy to move it into `mandatory_inlining_and_combine.sil` (no sure which is best).

zoecarver committed:
Remove unused ClosureCleanup from mandatory inlining

futurejones commented:
Can one of the admins verify this patch?

zoecarver committed:
Merge branch 'master' into optimize/dead-closure-in-mand-combine

zoecarver committed:
Merge branch 'master' into optimize/dead-closure-in-mand-combine

zoecarver committed:
Update instruction deleter callback to actually erease instructions so that cleanupLoadedCalleeValue can work properly (and other small fixes)

zoecarver committed:
Update mand inline sil tests after breaking closure cleanup into mand combine

zoecarver committed:
Removed unused member: tryRemoveUnused

zoecarver committed:
Fix remaining tests

zoecarver committed:
Run formatter over code

zoecarver commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1449,7 +1449,8 @@ static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
   if (!vla.computeFrontier(partialApplyFrontier,
-                           ValueLifetimeAnalysis::DontModifyCFG)) {
+                           ValueLifetimeAnalysis::DontModifyCFG) ||
+      partialApplyFrontier.empty()) {
```
If `partialApplyFrontier` is empty (and we don't return), we will create an `alloc_stack` and `copy_addr` but no `dealloc_stack`.

zoecarver commented:
@slavapestov and @gottesmm friendly ping. I've made sure that all the tests will pass after this patch is applied. Some of the tests appear to have been deleted but that either reflects an increase in deleted `partial_apply`s or they have been moved elsewhere (such as from `mandatory_inlining.sil` to `mandatory_inlining_and_combine.sil`). 

I realize this is a large patch so it may take a while to review. Let me know if you have any questions.

zoecarver committed:
Fix still failing tests

zoecarver commented:
@swift-ci please test.

zoecarver committed:
Merge branch 'master' into optimize/dead-closure-in-mand-combine

zoecarver commented:
@swift-ci please test.

gottesmm commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1943,3 +1944,102 @@ AbstractFunctionDecl *swift::getBaseMethod(AbstractFunctionDecl *FD) {
 }
+
+SILValue swift::stripCopiesAndBorrows(SILValue v) {
+  while (isa<CopyValueInst>(v) || isa<BeginBorrowInst>(v)) {
```
I think we already have a routine like this. I would look around.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -201,11 +203,6 @@ bool MandatoryCombiner::doOneIteration(SILFunction &function,
       );
     }
 
-    for (SILInstruction *instruction : instructionsPendingDeletion) {
```
Why is this removed?

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -284,6 +281,86 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
+    }
+  }
+
+  if (auto *convFunc = dyn_cast<ConvertFunctionInst>(val)) {
```
This is not needed. Should happen automatically when we go back around.

gottesmm reviewed (commented):
Some quick comments. I am going to want to relook at this after you make improvements/what we talked about over chat.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -284,6 +281,86 @@ SILInstruction *MandatoryCombiner::visitApplyInst(ApplyInst *instruction) {
+      }
+    }
+  } else if (auto *tttf = dyn_cast<ThinToThickFunctionInst>(src)) {
+    if (tryDeleteDeadClosure(tttf, instModCallbacks)) {
```
Most of the optimizations here should not be on LoadInst. You aren't removing load inst anywhere! Instead, all of these should be on the actual instructions you are removing (i.e. ThinToThick, Partial, Etc).

zoecarver committed:
Update based on review:
    * remove mand inlining changes
    * update combiner to be more like a "combiner"
    * delete trivially dead instructions
    * promote loads
    * don't delete instructions in the callback

zoecarver committed:
Reset tests / mand inline changes to master

zoecarver commented:
I'll make all these changes as their own (smaller, more focused) PRs.
